block-layer/codecs/dag-pb.md

```diff
@@ -44,13 +44,13 @@ This layout can be expressed with [IPLD Schemas](../../schemas/README.md) as:
 
 ```ipldsch
 type PBLink struct {
-  Hash Bytes
-  Name optional String
-  Tsize optional Int
+  Hash optional Link
+  Name String (implicit "")
```

Comment:
https://github.com/ipfs/go-merkledag/blob/749fd8717d46b4f34c9ce08253070079c89bc56d/pb/merkledag.pb.go#L90-L95

can't have empty string as far as I can tell, round-trip will end up with a `""`

Reply:
You can have an empty string. The type of `m.Name` is `*string`, so it can be `nil`, `""`, or any non-empty string. The code checks for `nil`, so you just end up with any string, empty or non-empty.

Reply:
Or maybe I misunderstand what you mean here? :)

Reply:
Yeah, this might need some experimental code! I have a bit of "made with Go" in my JS PB tests but maybe more would be good to sort out these specific pieces. There's really two questions I'm interested in:

1. What are the limitations on the IPLD nodes you can create for DAG-PB using go-ipld-prime-proto - can you make one that has a `nil` string as well as a `""` string and a `"foo"` string in the `Name` field?
2. What happens when you encode each of the above cases into protobufs and then decode it back into go-ipld-prime-proto; do you get all of these cases back out, `nil`, `""`, and `"foo"` or does `nil` always become `""` on such a round-trip as I'm suspecting.

And then as a follow-up - if there is variation between what you can _build_ in memory with go-ipld-prime-proto and what happens to it after you round-trip, is this acceptable? I'm taking the stance that this is not acceptable (at least for JS, but the reasoning should be the same for Go I think) because it leads to weird side-effects for users when their "nodes" are different after doing a round-trip through encode/decode. Same question for all of these fields that could be `optional` or have `implicit`s, same for the `Links` array or non-array.

I'm trying to be absolutely clear in JS and I think maybe I need to get dirty with go-ipld-prime-proto to do the same.